src/components/TypingHeading.tsx

```diff
@@ -1,48 +1,44 @@
 import classNames from "classnames";
-import { useCallback, useEffect, useState } from "react";
+import { useEffect, useState } from "react";
 import styles from "./TypingHeading.module.scss";
 
 export const TypingHeading = () => {
   const finalText = "we.wantToSleep()";
-  const [index, setIndex] = useState(1);
-  const [timerID, setTimerID] = useState<NodeJS.Timeout>();
+  // index of first highlighted character
+  const hlStart = 3;
+  // index of first non-highlighted character
+  const hlEnd = 14;
 
-  const [underscoreHidden, setUnderscoreHidden] = useState(false);
+  const [index, setIndex] = useState(1);
+  const [caretShown, setCaretShown] = useState(false);
 
   useEffect(() => {
-    setUnderscoreHidden(index % 14 < 7 && index > finalText.length);
-  }, [index]);
-
-  //   useEffect(() => {
-  //     if (index > finalText.length && timerID) {
-  //       clearInterval(timerID);
-  //     }
-  //   }, [index]);
+    const countdownID = setInterval(() => {
+      setIndex((prev) => {
+        if (prev === finalText.length) {
+          clearInterval(countdownID);
+          setCaretShown(true);
+        }
+        return prev + 1;
+      });
+    }, 100);
 
-  useEffect(() => {
-    const countdownID = setInterval(() => setIndex((prev) => prev + 1), 100);
-    setTimerID(countdownID.ref);
-    return () => {
-      clearInterval(countdownID);
-    };
+    return () => clearInterval(countdownID);
   }, []);
 
+  const h1ClassNames = classNames(
+    styles.heading,
+    caretShown && styles.blinkingCaret
```

Comment:
The recomended syntax would be:
classNames(styles.heading, {[styles.blinkingCaret]: caretShown} )
=> https://github.com/JedWatson/classnames

And for a better readability, I would put this line directly into the tsx